### ISSUE TITLE
Memory not being released in time, running out of memory

### ISSUE THREAD

**Issue by genotrance** [nim-lang/Nim#6031]
I'm using module securehash to generate hashes for files in order to identify duplicate files and am running out of memory since memory allocated by readFile() used in secureHashFile() is not being released in time. Invoking GC_fullCollect() within the loop doesn't help release memory no longer referenced.

See attached program [test.txt](https://github.com/nim-lang/Nim/files/1113148/test.txt), rename to .nim, simplified version which simply uses readFile() recursively on a directory. Ideally, it should be pointed to a directory with large files like videos. Observing memory usage in task manager shows nonstop increase. Nim quickly runs out of memory and exits on my laptop after hitting 1.6GB of RAM. In the loop, I have no need for the readFile() contents to stay in memory since content is not being referenced outside code block.

If pointed to a directory with many hundreds of smaller files (C:\Windows for example), Nim seems to manage without exiting but can still see memory usage not going down. On larger files, 100+MB each, the numbers add up quick and out of memory can be seen.

Running Nim 32-bit compiled from HEAD on Windows 10 64-bit.

Nim Compiler Version 0.17.0 (2017-06-29) [Windows: i386]
Copyright (c) 2006-2017 by Andreas Rumpf

git hash: 71c602122ccddb5d7b3360b8c5010488b2f58305
active boot switches: -d:release

**Comment by Araq** [comment]
Well but they are still in the stack which is not cleaned up because of your recursion. I'm afraid there is little we can do here. Remove the recursion and things should go well.

**Comment by genotrance** [comment]
Attaching a variation of the code where the readFile() is in a child function and the string content never makes it into the main loop. In theory, the readFile() content shouldn't be in the stack as far as recursion is concerned since the function that calls it returns.

[test2.txt](https://github.com/nim-lang/Nim/files/1114072/test2.txt)

Adding for the benefit of Varriount.

**Comment by andreaferretti** [comment]
The content of test2 is

```nim
import os

proc doread(file: string) =
    var p = readFile(file)
    echo file & " " & $p.len()

proc recurse(dir: string) =
    for f in walkPattern(dir & DirSep & "/*"):
        try:
            let info = getFileInfo(f)
            if info.kind == pcFile:
                doread(f)
            elif info.kind == pcDir:
                recurse(f)
        except:
            continue

recurse("c:/windows")
```

**Comment by yglukhov** [comment]
Might be related: https://github.com/nim-lang/Nim/issues/5589

**Comment by genotrance** [comment]
I tested the script with v0.17.2 stable as a baseline and was able to reproduce the issue on my Windows directory of 30GB / 140k files. It went up to 1.6GB of RAM before running out of memory.

I then tested with Nim devel and it didn't fail, although it did go upto 1.55GB of RAM usage. I'm not sure if it's fixed but at least it didn't run out of memory for this data set.

```
Nim Compiler Version 0.17.3 [Windows: i386]
Copyright (c) 2006-2017 by Andreas Rumpf

git hash: 28e0bf9dcd62f387c79f767848cadd8b71d825de
active boot switches: -d:release
```
